### DISCUSSION TITLE
YOLO model make threads call dummy - d%

### DISCUSSION THREAD

**Discussion by mr-mahmood** [ultralytics/ultralytics#19076] [Q&A] [answered]
so i am working on a project and i notice somthing strange when i run my code and look at threads that are working i see some threads pop up once a while and when i delete part of code that is related to YOLO model threads stop poping up and i dont have any problem now, but i want to use yolo, so i wonder why is this happening and is it effect performance or no because i notice after some hours it make a lot of dummy threads that all of them are running and it seems like it must have effects
![Screenshot 2025-02-05 033334](https://github.com/user-attachments/assets/707c13eb-2e04-493c-bad1-ea0dceaec3e4)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @mr-mahmood, thank you for reaching out and bringing this to our attention 🚀! We recommend checking out the [Docs](https://docs.ultralytics.com/) as a starting point to explore YOLO usage, as many questions are addressed there. You can find detailed [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples. Additionally, you can review the [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) to validate your implementation.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/) to clarify the issue. This will help us debug it more effectively.

Based on your description, it could be related to threading or resource management within the YOLO model. If possible:
1. Share a code snippet highlighting how you're integrating the YOLO model.
2. Specify your environment (OS, Python version, PyTorch version, etc.).
3. Confirm whether you've tested this on the latest Ultralytics version by upgrading with:

```bash
pip install -U ultralytics
```

You may also want to monitor system resources and verify your setup aligns with the hardware and software requirements listed in our [verified environments](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml). To test YOLO on free cloud GPUs, use one of the following platforms:  
- **Google Colab**: <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a>  
- **Paperspace Gradient**: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a>  
- **Kaggle**: <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>  

For broader discussions or technical collaboration, join the Ultralytics community:  
- [Discord](https://discord.com/invite/ultralytics) 🎧 for real-time support  
- [Discourse](https://community.ultralytics.com/) for structured topic-based discussions  
- [Subreddit](https://reddit.com/r/Ultralytics) for community insights  

Ensure that you are using an environment that meets all dependencies (e.g., [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn)). If the issue persists after these steps, feel free to elaborate further, and an Ultralytics engineer will assist you soon! 😊 

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If the CI badge above is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing, and your issue may be specific to your setup.  

We appreciate your patience and look forward to resolving this! 🚀

**Reply by mr-mahmood** [reply]
hi, thanks for respond, i use newest version of all library and my python version is 3.13.1, and code that i write is something like this:

```python

import socket
import time
import struct
import numpy as np
import json
import cv2
import threading
import select  # For efficient waiting on UDP socket
from multiprocessing.shared_memory import SharedMemory
from multiprocessing import Event  # You can also use threading.Event if preferred
import torch
from ultralytics import YOLO

class FrameConsumer:
    def __init__(self, camera_name, udp_ip, udp_port, shm_name, buffer_size=10, width=480, height=640):
        # Use a thread-safe event for stopping the thread
        self.stop_event = Event()
        
        # Set device to GPU if available, otherwise CPU
        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
        self.model1 = YOLO('./AI_model/models/Full license plate detection model.pt').to(self.device)
        self.model2 = YOLO('./AI_model/models/farsi characters detection model.pt').to(self.device)
        
        self.camera_name = camera_name
        self.udp_ip = udp_ip
        self.udp_port = udp_port
        
        # Connect to existing shared memory
        self.shm = SharedMemory(name=shm_name)
        self.buffer_size = buffer_size
        
        self.last_write_idx = -1  # To track the last processed frame
        
        # Read header metadata (6 integers = 24 bytes)
        self.header_data = struct.unpack("IIIIII", self.shm.buf[0:24])
        
        self.width = width
        self.height = height
        self.channels = 3
            
        self.header_size = 24  # 6 integers * 4 bytes each
        self.frame_size = self.width * self.height * self.channels
        
        # Setup UDP socket in non-blocking mode
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self.sock.bind((self.udp_ip, self.udp_port))
        self.sock.setblocking(0)
        
        self.latest_data = None
        self.lock = threading.Lock()  # For thread-safe access to latest_data
    
        # Start the UDP listener thread
        self.udp_thread = threading.Thread(name='udp_listener', target=self._udp_listener, daemon=True)
        self.udp_thread.start()

    def _udp_listener(self):
        """ Continuously listen for UDP packets and update the latest data. """
        while not self.stop_event.is_set():
            # Use select to wait until the socket is ready for reading
            rlist, _, _ = select.select([self.sock], [], [], 0.005)
            if rlist:
                try:
                    data, _ = self.sock.recvfrom(1024)  # Read UDP packet (should be 16 bytes for "IId")
                    with self.lock:
                        self.latest_data = data  # Save the latest UDP packet
                except Exception as e:
                    # You might log this exception in production
                    continue
            else:
                continue

    def _get_frame(self):
        """ Retrieve the latest frame based on the most recent UDP packet. """
        with self.lock:
            latest_data = self.latest_data
            self.latest_data = None  # Clear the stored UDP data

        if latest_data is None:
            return None, None  # No new data available

        # Unpack the UDP packet: write index, slot index, and send time
        write_idx, slot_idx, send_time = struct.unpack("IId", latest_data)

        # Only process new frames
        if write_idx <= self.last_write_idx:
            return None, None

        self.last_write_idx = write_idx

        # Calculate offset into the shared memory and extract the frame data
        offset = self.header_size + (slot_idx * self.frame_size)
        frame_buf = self.shm.buf[offset:offset+self.frame_size]
        frame = np.frombuffer(frame_buf, dtype=np.uint8).reshape((self.height, self.width, self.channels))

        return frame, send_time
              
    def run(self):
        try:
            while not self.stop_event.is_set():
                frame, send_time = self._get_frame()
                if frame is None:
                    continue

                # Calculate latency before processing
                pre_latency = time.time() - send_time
                cv2.putText(frame, f'latency before model = {pre_latency*1000:.0f} ms', (5, 20),
                            cv2.FONT_HERSHEY_SIMPLEX, 0.4, (255, 255, 0), 1, cv2.LINE_AA)

                # Run first model for license plate detection
                results = self.model1(frame, verbose=False)[0]
                
                for result in results.boxes.data.tolist():
                    x1, y1, x2, y2, score, class_id = result
                    if score > 0.5:
                        # Crop the detected region with a small margin (adjust if needed)
                        crop_y1 = max(int(y1) - 35, 0)
                        cropped_frame = frame[crop_y1:int(y2), int(x1):int(x2)]
                        cv2.rectangle(frame, (int(x1), int(y1)), (int(x2), int(y2)), (255, 0, 255), 1)

                        # Run second model for character recognition
                        results2 = self.model2(cropped_frame, verbose=False)[0]
                        
                        for result2 in results2.boxes.data.tolist():
                            cx1, cy1, cx2, cy2, score2, class_id2 = result2
                            if score2 > 0.5:
                                label = results2.names.get(int(class_id2), str(class_id2))
                                cv2.putText(cropped_frame, label, (int(cx1), 25),
                                            cv2.FONT_HERSHEY_SIMPLEX, 0.8, (0, 255, 0), 2, cv2.LINE_AA)

                # Calculate latency after processing
                post_latency = time.time() - send_time
                cv2.putText(frame, f'latency after model = {post_latency*1000:.0f} ms', (5, 35),
                            cv2.FONT_HERSHEY_SIMPLEX, 0.4, (255, 255, 0), 1, cv2.LINE_AA)

                # Display the frame
                cv2.imshow(self.camera_name, frame)
                if cv2.waitKey(1) & 0xFF == ord('q'):
                    self.stop_event.set()
                    break

        except KeyboardInterrupt:
            # Allow graceful interruption with Ctrl+C
            self.stop_event.set()

        finally:
            # Cleanup resources
            self.stop_event.set()
            if self.udp_thread.is_alive():
                self.udp_thread.join()
            self.sock.close()
            cv2.destroyAllWindows()

if __name__ == "__main__":
    # Load camera configuration from JSON file
    with open('./AI_model/Layer-by-layer implementation/cameras.json', 'r', encoding='utf-8') as f:
        config = json.load(f)
    
    index = 0
    cam = config['cameras'][index]
    consumer = FrameConsumer(
        camera_name=cam['name'],
        udp_ip=cam['udp_ip'],
        udp_port=cam['udp_port'],
        shm_name=cam['shm_name'],
        buffer_size=cam['buffer_size'],
        width=cam['width'],
        height=cam['height']
    )
        
    consumer.run()

```

**Reply by glenn-jocher** [reply, marked as answer]
@mr-mahmood the dummy threads you're observing are part of YOLO's normal video streaming pipeline for efficient frame processing, as shown in our LoadStreams implementation. These daemon threads (marked with `daemon=True` in the code) are automatically managed and shouldn't impact performance when properly cleaned up. To validate, you could:

1. Profile GPU/CPU utilization during extended runs
2. Add explicit cleanup in your `finally` block:
```python
del self.model1, self.model2
torch.cuda.empty_cache()
```

If performance degradation persists, please share a minimal reproducible example focusing just on the YOLO inference loop.
